### ISSUE TITLE
DotNetObjectReference multiple instances created, only last one disposed

### ISSUE THREAD

**Issue by omuleanu** [dotnet/aspnetcore#44064]
I see in docs you have code that only disposes the last instance of `DotNetObjectReference`
```
@page "/call-dotnet-example-2"
...
    <button @onclick="TriggerDotNetInstanceMethod">
        Trigger .NET instance method
...
@code {
  ...
    public async Task TriggerDotNetInstanceMethod()
    {
        objRef = DotNetObjectReference.Create(this);
        result = await JS.InvokeAsync<string>("sayHello1", objRef);
    }
    public void Dispose()
    {
        objRef?.Dispose();
    }
}
```
You're calling `DotNetObjectReference.Create` on each button click, so `Dispose` will only execute on the last created instance. 
Is that an issue ?

---
also I see here: 
https://docs.microsoft.com/en-us/dotnet/api/microsoft.jsinterop.dotnetobjectreference-1?view=aspnetcore-6.0

> the reference must later be disposed by JS code or by .NET code

can you dispose the DotNetObjectReference in JS or it is an error in the doc ?

cc: @guardrex https://github.com/dotnet/AspNetCore.Docs/issues/27036

### .NET Version

6

**Comment by guardrex** [comment]
> an you dispose the DotNetObjectReference in JS

See the end of ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?view=aspnetcore-6.0#class-instance-examples

... and the example at ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?view=aspnetcore-6.0#component-instance-net-method-helper-class

**Comment by omuleanu** [comment]
@mkArtakMSFT the main question was not answered.

> You're calling DotNetObjectReference.Create on each button click, so Dispose will only execute on the last created instance.
Is that an issue ?

**Comment by guardrex** [comment]
That's correct @mkArtakMSFT ... I was only addressing one part of @omuleanu's remarks.

My understanding per docs PR feedback (e.g., Pranav) is that calling `DotNetObjectReference.Create(this)` does not create new, instances that need to be tracked for individual disposal. @omuleanu's question stands, and we probably should add a remark to the topic on this point after I hear back 👂 what's going on under-the-covers.

**Comment by SteveSandersonMS** [comment]
I verified that, as expected, `DotNetObjectReference.Create` returns different instances on each call, and hence each instance must be disposed separately.

**Comment by guardrex** [comment]
Very well. I'll move the assignment to component initialization so that it's only called once.

**Comment by SteveSandersonMS** [comment]
Not blaming you, @guardrex! Thanks for updating the docs.

**Comment by omuleanu** [comment]
@guardrex think you could initialize it on the first click also

**Comment by guardrex** [comment]
@MackinnonBuck @SteveSandersonMS ... Does the following use (no assignment) still require me to create the object reference in init, or is the following ok?

```razor
@code {
    public async Task TriggerDotNetInstanceMethod()
    {
        result = await JS.InvokeAsync<string>("sayHello1", 
            DotNetObjectReference.Create(this));
    }
}
```

**Comment by omuleanu** [comment]
@guardrex think you need to assign it so you could dispose it

**Comment by omuleanu** [comment]
```
@code {
...
    public async Task TriggerDotNetInstanceMethod()
    {
        if (objRef is null)
        {
            objRef = DotNetObjectReference.Create(this);
        }

        result = await JS.InvokeAsync<string>("sayHello1", objRef);
    }

    public void Dispose()
    {
        objRef?.Dispose();
    }
}
```

**Comment by guardrex** [comment]
Yes, I'm aware of that, but I think we have an example somewhere that doesn't require disposal when it isn't assigned. It will simplify a number of our examples if I can use the same pattern.

**Comment by MackinnonBuck** [comment]
@guardrex The `DotNetObjectReference` has to be explicitly disposed at some point. If it's passed directly as a parameter to a JS invocation like that, then the JS method must dispose it. Otherwise, you need to assign the `DotNetObjectReference` to some field or local and dispose of it explicitly in .NET.

**Comment by guardrex** [comment]
> JS method must dispose it

Ah! Yes! I remember now that you said that. The example that I'm recalling probably does do that, and I will check that as part of my pass over all of these examples.

I'll take care of this further from the docs issue/PR.

There's one tricky one where a class instance takes a parameter in it's constructor that is, like these other examples, subject to button click events creating the instances and thus having the same problem. I'll check and see if that one is disposing on the JS side. If not, I'll ping u on the docs issue to see what we need to do with it if I can't figure it out.
